patroni/ha.py

```diff
@@ -293,54 +294,103 @@ def is_synchronous_mode(self):
-                    logger.info("Removing synchronous privilege from %s", current)
-                    if not self.dcs.write_sync_state(self.state_handler.name, None, index=self.cluster.sync.index):
-                        logger.info('Synchronous replication key updated by someone else.')
-                        return
+            sync_state = self.state_handler.current_sync_state(self.cluster)
 
-                if self.is_synchronous_mode_strict() and picked is None:
-                    picked = '*'
-                    logger.warning("No standbys available!")
-
-                logger.info("Assigning synchronous standby status to %s", picked)
-                self.state_handler.set_synchronous_standby(picked)
-
-                if picked and picked != '*' and not allow_promote:
-                    # Wait for PostgreSQL to enable synchronous mode and see if we can immediately set sync_standby
-                    time.sleep(2)
-                    picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
-                if allow_promote:
-                    cluster = self.dcs.get_cluster()
-                    if cluster.sync.leader and cluster.sync.leader != self.state_handler.name:
-                        logger.info("Synchronous replication key updated by someone else")
-                        return
-                    if not self.dcs.write_sync_state(self.state_handler.name, picked, index=cluster.sync.index):
-                        logger.info("Synchronous replication key updated by someone else")
+            min_sync = self.patroni.config['minimum_replication_factor']
+            sync_wanted = max(self.patroni.config['replication_factor'], min_sync)
+
+            # Versions before 9.6 only support synchronous replication to one node
+            if not self.state_handler.use_multiple_sync:
+                sync_wanted = min(sync_wanted, 2)
+                min_sync = min(min_sync, 2)
+
+            if self.cluster.sync.is_empty:
+                quorum, voters = 1, frozenset([self.state_handler.name])
+            else:
+                quorum, voters = self.cluster.sync.quorum, self.cluster.sync.members
+
+            for transition, num, nodes in QuorumStateResolver(quorum=quorum,
+                                                              voters=voters,
+                                                              numsync=sync_state['numsync'],
+                                                              sync=sync_state['sync'],
+                                                              active=sync_state['active'],
+                                                              sync_wanted=sync_wanted):
+                if transition == 'quorum':
+                    logger.info("Setting quorum to %d of %d (%s)", num, len(nodes), ", ".join(sorted(nodes)))
+                    if not self.dcs.write_sync_state(leader=self.state_handler.name, quorum=num, members=list(nodes),
```

Comment:
`write_sync_state` - this method should be also changed, no ?
here you are passing `leader`, `quorum`, `membres` and `index` as a parameters, but signature of method wasn't changed in this pull request.

```python
def write_sync_state(self, leader, sync_standby, index=None):
```